File: src/sections/gitstatus.js
const { execFile } = require('child_process');
const fs = require('fs');
const path = require('path');

function run(cmd, args, opts = {}) {
  return new Promise((resolve) => {
    execFile(cmd, args, { timeout: 10000, ...opts }, (err, stdout, stderr) => {
      if (err) resolve({ error: err.message, stdout: '', stderr });
      else resolve({ stdout: stdout.trim(), stderr: stderr.trim() });
    });
  });
}

async function scanRepo(repoPath) {
  const name = path.basename(repoPath);
  const cwd = repoPath;

  // Check if it's a git repo
  const gitDir = path.join(repoPath, '.git');
  try {
    fs.accessSync(gitDir);
  } catch {
    return null; // Not a git repo
  }

  const info = { name, path: repoPath };

  // Uncommitted changes
  const status = await run('git', ['status', '--porcelain'], { cwd });
  if (!status.error && status.stdout) {
    const changes = status.stdout.split('\n').filter(l => l.trim());
    info.uncommitted = changes.length;
  } else {
    info.uncommitted = 0;
  }

  // Current branch
  const branch = await run('git', ['branch', '--show-current'], { cwd });
  info.branch = branch.error ? 'unknown' : (branch.stdout || 'detached');

  // Ahead/behind
  const ab = await run('git', ['rev-list', '--left-right', '--count', `HEAD...@{upstream}`], { cwd });
  if (!ab.error && ab.stdout) {
    const parts = ab.stdout.split(/\s+/);
    info.ahead = parseInt(parts[0]) || 0;
    info.behind = parseInt(parts[1]) || 0;
  }

  // Recent commits (last 24h)
  const since = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString();
  const log = await run('git', ['log', `--since=${since}`, '--oneline', '--no-merges'], { cwd });
  if (!log.error && log.stdout) {
    info.recentCommits = log.stdout.split('\n').filter(l => l.trim()).length;
  } else {
    info.recentCommits = 0;
  }

  return info;
}

async function getGitStatus(args) {
  const repos = [];
  const errors = [];

  for (const dir of args.gitDirs) {
    const expandedDir = dir.replace(/^~/, require('os').homedir());
    try {
      const entries = fs.readdirSync(expandedDir, { withFileTypes: true });
      const subdirs = entries
        .filter(e => e.isDirectory() && !e.name.startsWith('.'))
        .map(e => path.join(expandedDir, e.name));

      const results = await Promise.all(subdirs.map(scanRepo));
      repos.push(...results.filter(Boolean));
    } catch (err) {
      errors.push(`${dir}: ${err.message}`);
    }
  }

  // Sort: repos with changes first, then by name
  repos.sort((a, b) => {
    const aActive = (a.uncommitted > 0 || a.ahead > 0) ? 1 : 0;
    const bActive = (b.uncommitted > 0 || b.ahead > 0) ? 1 : 0;
    if (bActive !== aActive) return bActive - aActive;
    return a.name.localeCompare(b.name);
  });

  const dirty = repos.filter(r => r.uncommitted > 0);
  const withRecent = repos.filter(r => r.recentCommits > 0);

  return {
    totalRepos: repos.length,
    dirtyRepos: dirty.length,
    reposWithRecentCommits: withRecent.length,
    repos,
    errors: errors.length ? errors : undefined,
  };
}

module.exports = { getGitStatus };

File: daily_briefing/sections/gitstatus.py
"""Git status section — scans repos for uncommitted changes, ahead/behind, recent commits."""

import os
import subprocess
from datetime import datetime, timedelta, timezone
from pathlib import Path


def _run(cmd, cwd=None, timeout=10):
    """Run a command and return (stdout, error)."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
            cwd=cwd,
        )
        if result.returncode != 0:
            return "", result.stderr.strip()
        return result.stdout.strip(), None
    except (subprocess.TimeoutExpired, FileNotFoundError) as e:
        return "", str(e)


def _scan_repo(repo_path):
    """Scan a single git repo for status info."""
    git_dir = repo_path / ".git"
    if not git_dir.exists():
        return None

    name = repo_path.name
    info = {"name": name, "path": str(repo_path)}
    cwd = str(repo_path)

    # Uncommitted changes
    out, err = _run(["git", "status", "--porcelain"], cwd=cwd)
    if err is None and out:
        info["uncommitted"] = len([l for l in out.splitlines() if l.strip()])
    else:
        info["uncommitted"] = 0

    # Current branch
    out, err = _run(["git", "branch", "--show-current"], cwd=cwd)
    info["branch"] = out if not err else "unknown"
    if not info["branch"]:
        info["branch"] = "detached"

    # Ahead/behind
    out, err = _run(
        ["git", "rev-list", "--left-right", "--count", "HEAD...@{upstream}"],
        cwd=cwd,
    )
    if not err and out:
        parts = out.split()
        info["ahead"] = int(parts[0]) if len(parts) > 0 else 0
        info["behind"] = int(parts[1]) if len(parts) > 1 else 0
    else:
        info["ahead"] = 0
        info["behind"] = 0

    # Recent commits (last 24h)
    since = (datetime.now(timezone.utc) - timedelta(days=1)).isoformat()
    out, err = _run(
        ["git", "log", f"--since={since}", "--oneline", "--no-merges"],
        cwd=cwd,
    )
    if not err and out:
        info["recentCommits"] = len([l for l in out.splitlines() if l.strip()])
    else:
        info["recentCommits"] = 0

    return info


def get_git_status(args):
    repos = []
    errors = []

    for d in args.git_dirs:
        expanded = Path(d).expanduser()
        try:
            entries = sorted(expanded.iterdir())
            for entry in entries:
                if entry.is_dir() and not entry.name.startswith("."):
                    result = _scan_repo(entry)
                    if result:
                        repos.append(result)
        except OSError as e:
            errors.append(f"{d}: {e}")

    # Sort: repos with activity first, then alphabetical
    repos.sort(
        key=lambda r: (
            -(1 if r["uncommitted"] > 0 or r.get("ahead", 0) > 0 else 0),
            r["name"].lower(),
        )
    )

    dirty = [r for r in repos if r["uncommitted"] > 0]
    with_recent = [r for r in repos if r["recentCommits"] > 0]

    result = {
        "totalRepos": len(repos),
        "dirtyRepos": len(dirty),
        "reposWithRecentCommits": len(with_recent),
        "repos": repos,
    }
    if errors:
        result["errors"] = errors

    return result
